File: src/renderer/utils/logger.ts
const DEFAULT_STRING_MAX_LEN = 400;
const DEFAULT_MULTILINE_MAX_LEN = 2000;
const DEFAULT_MAX_LINES = 30;

export const truncate = (value: string, maxLen = DEFAULT_STRING_MAX_LEN) => {
  const text = typeof value === "string" ? value : String(value);
  return text.length > maxLen
    ? `${text.slice(0, maxLen)}...[truncated ${text.length - maxLen} chars]`
    : text;
};

export const truncateLines = (
  value: string,
  maxLen = DEFAULT_MULTILINE_MAX_LEN,
  maxLines = DEFAULT_MAX_LINES,
) => {
  const text = typeof value === "string" ? value : String(value);
  const lines = text.split(/\r?\n/);
  const visibleLines = lines.slice(0, maxLines);
  let preview = visibleLines.join("\n");

  if (preview.length > maxLen) {
    preview = preview.slice(0, maxLen);
  }

  const hiddenChars = Math.max(0, text.length - preview.length);
  const hiddenLines = Math.max(0, lines.length - visibleLines.length);

  if (hiddenChars === 0 && hiddenLines === 0) {
    return preview;
  }

  const markers: string[] = [];
  if (hiddenLines > 0) {
    markers.push(`${hiddenLines} lines`);
  }
  if (hiddenChars > 0) {
    markers.push(`${hiddenChars} chars`);
  }

  return `${preview}\n...[truncated ${markers.join(", ")}]`;
};

const serializeError = (error: Error) => ({
  name: error.name,
  message: error.message,
  stack: truncateLines(error.stack ?? error.message, 3000, 40),
});

const safeStringify = (value: unknown) => {
  const seen = new WeakSet<object>();

  try {
    return truncateLines(
      JSON.stringify(
        value,
        (_key, currentValue: unknown) => {
          if (typeof currentValue === "string") {
            return truncateLines(currentValue, DEFAULT_STRING_MAX_LEN, 10);
          }
          if (currentValue instanceof Error) {
            return serializeError(currentValue);
          }
          if (
            currentValue &&
            typeof currentValue === "object" &&
            !Array.isArray(currentValue)
          ) {
            if (seen.has(currentValue)) {
              return "[Circular]";
            }
            seen.add(currentValue);
          }
          return currentValue;
        },
        2,
      ),
      DEFAULT_MULTILINE_MAX_LEN,
      DEFAULT_MAX_LINES,
    );
  } catch {
    return "[JSON serialization failed]";
  }
};

const formatArg = (arg: unknown) => {
  if (arg instanceof Error) {
    return safeStringify(serializeError(arg));
  }
  if (typeof arg === "string") {
    return truncateLines(arg);
  }
  if (
    typeof arg === "number" ||
    typeof arg === "boolean" ||
    arg === null ||
    arg === undefined
  ) {
    return String(arg);
  }
  return safeStringify(arg);
};

const timestamp = () => new Date().toISOString();

const emit = (
  method: "debug" | "info" | "warn" | "error",
  level: string,
  tag: string,
  args: unknown[],
) => {
  console[method](`[${timestamp()}] [${level}] [${tag}]`, ...args.map(formatArg));
};

export const createRendererLogger = (tag: string) => ({
  DEBUG: (...args: unknown[]) => emit("debug", "DEBUG", tag, args),
  INFO: (...args: unknown[]) => emit("info", "INFO", tag, args),
  WARN: (...args: unknown[]) => emit("warn", "WARN", tag, args),
  ERROR: (...args: unknown[]) => emit("error", "ERROR", tag, args),
});

File: python/logger.py
import json
import traceback
from contextvars import ContextVar
from datetime import datetime, timezone
from typing import Any, Mapping

from colorama import Fore, Style, init

init(autoreset=True)

_LOG_CONTEXT: ContextVar[dict[str, Any]] = ContextVar("_LOG_CONTEXT", default={})

_LEVEL_COLORS = {
    "DEBUG": Fore.CYAN,
    "INFO": Fore.BLUE,
    "SUCCESS": Fore.GREEN,
    "WARNING": Fore.YELLOW,
    "ERROR": Fore.RED,
}


def truncate_text(value: Any, max_len: int = 600, max_lines: int = 20) -> str:
    text = value if isinstance(value, str) else str(value)
    lines = text.splitlines()
    visible_lines = lines[:max_lines]
    preview = "\n".join(visible_lines)

    if len(preview) > max_len:
        preview = preview[:max_len]

    hidden_chars = max(0, len(text) - len(preview))
    hidden_lines = max(0, len(lines) - len(visible_lines))

    if hidden_chars == 0 and hidden_lines == 0:
        return preview

    markers: list[str] = []
    if hidden_lines > 0:
        markers.append(f"{hidden_lines} lines")
    if hidden_chars > 0:
        markers.append(f"{hidden_chars} chars")

    return f"{preview}\n...[truncated {', '.join(markers)}]"


def _serialize_value(value: Any) -> str:
    if isinstance(value, BaseException):
        return json.dumps(
            {
                "name": type(value).__name__,
                "message": str(value),
                "stack": truncate_text(
                    "".join(
                        traceback.format_exception(
                            type(value), value, value.__traceback__
                        )
                    ),
                    max_len=3000,
                    max_lines=40,
                ),
            },
            ensure_ascii=True,
        )

    if isinstance(value, (dict, list, tuple, set)):
        try:
            return truncate_text(
                json.dumps(value, default=str, ensure_ascii=True),
                max_len=800,
                max_lines=20,
            )
        except Exception:
            return truncate_text(value)

    return truncate_text(value, max_len=240, max_lines=8)


def set_log_context(**context: Any) -> None:
    current = dict(_LOG_CONTEXT.get())
    for key, value in context.items():
        if value is None:
            current.pop(key, None)
        else:
            current[key] = value
    _LOG_CONTEXT.set(current)


def clear_log_context(*keys: str) -> None:
    if not keys:
        _LOG_CONTEXT.set({})
        return

    current = dict(_LOG_CONTEXT.get())
    for key in keys:
        current.pop(key, None)
    _LOG_CONTEXT.set(current)


def _format_context(context: Mapping[str, Any] | None = None) -> str:
    merged = dict(_LOG_CONTEXT.get())
    if context:
        merged.update({key: value for key, value in context.items() if value is not None})

    if not merged:
        return ""

    parts = [f"{key}={_serialize_value(value)}" for key, value in merged.items()]
    return " | " + " ".join(parts)


def _log(
    level: str,
    message: str,
    *,
    context: Mapping[str, Any] | None = None,
    exc_info: BaseException | None = None,
) -> None:
    color = _LEVEL_COLORS[level]
    timestamp = datetime.now(timezone.utc).isoformat(timespec="milliseconds")
    context_suffix = _format_context(context)

    print(
        f"{color}{Style.BRIGHT}{timestamp} {level}: {truncate_text(message, max_len=500, max_lines=8)}{context_suffix}{Style.RESET_ALL}"
    )

    if exc_info is not None:
        stack = truncate_text(
            "".join(traceback.format_exception(type(exc_info), exc_info, exc_info.__traceback__)),
            max_len=4000,
            max_lines=60,
        )
        print(f"{color}{stack}{Style.RESET_ALL}")


def log_debug(message: str, context: Mapping[str, Any] | None = None) -> None:
    _log("DEBUG", message, context=context)


def log_error(
    message: str,
    context: Mapping[str, Any] | None = None,
    exc_info: BaseException | None = None,
) -> None:
    _log("ERROR", message, context=context, exc_info=exc_info)


def log_success(message: str, context: Mapping[str, Any] | None = None) -> None:
    _log("SUCCESS", message, context=context)


def log_info(message: str, context: Mapping[str, Any] | None = None) -> None:
    _log("INFO", message, context=context)


def log_warning(message: str, context: Mapping[str, Any] | None = None) -> None:
    _log("WARNING", message, context=context)